readium/shared/src/main/java/org/readium/r2/shared/resource/Resource.kt

```diff
@@ -0,0 +1,188 @@
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.nio.charset.Charset
+import org.json.JSONObject
+import org.readium.r2.shared.R
+import org.readium.r2.shared.UserException
+import org.readium.r2.shared.error.Try
+import org.readium.r2.shared.error.flatMap
+import org.readium.r2.shared.parser.xml.ElementNode
+import org.readium.r2.shared.parser.xml.XmlParser
+import org.readium.r2.shared.util.SuspendingCloseable
+
+typealias ResourceTry<SuccessT> = Try<SuccessT, Resource.Exception>
+
+/**
+ * Acts as a proxy to an actual resource by handling read access.
+ */
+interface Resource : SuspendingCloseable {
+
+    /**
+     * Direct file to this resource, when available.
+     *
+     * This is meant to be used as an optimization for consumers which can't work efficiently
+     * with streams. However, [file] is not guaranteed to be set, for example if the resource
+     * underwent transformations or is being read from an archive. Therefore, consumers should
+     * always fallback on regular stream reading, using [read] or [ResourceInputStream].
+     */
+    val file: File? get() = null
+
+    /**
+     * Returns the resource media type if known.
+     */
+    suspend fun mediaType(): ResourceTry<String?> = ResourceTry.success(null)
+
+    /**
+     * Returns the name of the resource if any.
+     */
+    suspend fun name(): ResourceTry<String?> = ResourceTry.success(null)
+
+    /**
+     * Returns data length from metadata if available, or calculated from reading the bytes otherwise.
+     *
+     * This value must be treated as a hint, as it might not reflect the actual bytes length. To get
+     * the real length, you need to read the whole resource.
+     */
+    suspend fun length(): ResourceTry<Long>
+
+    /**
+     * Reads the bytes at the given range.
+     *
+     * When [range] is null, the whole content is returned. Out-of-range indexes are clamped to the
+     * available length automatically.
+     */
+    suspend fun read(range: LongRange? = null): ResourceTry<ByteArray>
+
+    /**
+     * Reads the full content as a [String].
+     *
+     * If [charset] is null, then it is parsed from the `charset` parameter of link().type,
+     * or falls back on UTF-8.
+     */
+    suspend fun readAsString(charset: Charset? = null): ResourceTry<String> =
+        read().mapCatching {
+            String(it, charset = charset ?: Charsets.UTF_8)
+        }
+
+    /**
+     * Reads the full content as a JSON object.
+     */
+    suspend fun readAsJson(): ResourceTry<JSONObject> =
+        readAsString(charset = Charsets.UTF_8).mapCatching { JSONObject(it) }
+
+    /**
+     * Reads the full content as an XML document.
+     */
+    suspend fun readAsXml(): ResourceTry<ElementNode> =
+        read().mapCatching { XmlParser().parse(ByteArrayInputStream(it)) }
+
+    /**
+     * Reads the full content as a [Bitmap].
+     */
+    suspend fun readAsBitmap(): ResourceTry<Bitmap> =
+        read().mapCatching {
```

Comment:
I would implement these functions as extensions on `Resource`, to make it clear that they are not part of the interface contract but helpers.